ecs-agent/daemonimages/csidriver/driver/node_test.go

```diff
@@ -0,0 +1,161 @@
+//go:build linux && unit
+// +build linux,unit
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package driver
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/golang/mock/gomock"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	volumeID = "voltest"
+	nvmeName = "/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_voltest"
+)
+
+func TestNodeGetVolumeStats(t *testing.T) {
+	testCases := []struct {
+		name     string
+		testFunc func(t *testing.T)
+	}{
+		{
+			name: "success normal",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockMounter := NewMockMounter(mockCtl)
+				VolumePath := "./test"
+				err := os.MkdirAll(VolumePath, 0644)
+				if err != nil {
+					t.Fatalf("fail to create dir: %v", err)
+				}
+				defer os.RemoveAll(VolumePath)
+
+				mockMounter.EXPECT().PathExists(VolumePath).Return(true, nil)
+
+				awsDriver := nodeService{
+					mounter: mockMounter,
+				}
+
+				req := &csi.NodeGetVolumeStatsRequest{
+					VolumeId:   volumeID,
+					VolumePath: VolumePath,
+				}
+				_, err = awsDriver.NodeGetVolumeStats(context.TODO(), req)
+				if err != nil {
+					t.Fatalf("Expect no error but got: %v", err)
+				}
+			},
+		},
+		{
+			name: "fail path not exist",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockMounter := NewMockMounter(mockCtl)
+				VolumePath := "/test"
+
+				mockMounter.EXPECT().PathExists(VolumePath).Return(false, nil)
+
+				awsDriver := nodeService{
+					mounter: mockMounter,
+				}
+
+				req := &csi.NodeGetVolumeStatsRequest{
+					VolumeId:   volumeID,
+					VolumePath: VolumePath,
+				}
+				_, err := awsDriver.NodeGetVolumeStats(context.TODO(), req)
+				expectErr(t, err, codes.NotFound)
+			},
+		},
+		{
+			name: "fail can't determine block device",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockMounter := NewMockMounter(mockCtl)
+				VolumePath := "/test"
+
+				mockMounter.EXPECT().PathExists(VolumePath).Return(true, nil)
+
+				awsDriver := nodeService{
+					mounter: mockMounter,
+				}
+
+				req := &csi.NodeGetVolumeStatsRequest{
+					VolumeId:   volumeID,
+					VolumePath: VolumePath,
+				}
+				_, err := awsDriver.NodeGetVolumeStats(context.TODO(), req)
```

Comment:
Question: What error happens in this case? Permission denied or file/directory doesn't exist?

Reply:
It could be -
- VolumePath not exist
- Some system call failed. e.g. unable to get the stats of this path